### PR TITLE
App Banner Support for Android Chrome

### PR DESCRIPTION
**Pull request by xfalcox**
https://meta.discourse.org/t/fully-support-android-chromes-add-to-homescreen/21067/17?u=falco

### CODE REVIEW

**Comment by eviltrout** [comment]
Seems a bit wasteful to instantiate a service worker that does nothing just for one platform.

Perhaps it should only be spawned on android if it is only necessary on that platform?

**Comment by xfalcox** [comment]
That's a good idea @eviltrout :+1: 

There is user agent detection already set up somewhere client side?

Also, I'm reading Push Notification API for Chrome and Firefox on Android, we can re-purpose this service worker later.

**Comment by eviltrout** [comment]
We do have client side detection of android, but it won't be set up in the `_discourse_javascript` file.

Instead I suggest you create a new initializer. in the initializer you can ask the container for the capabilities object and check `isAndroid`:

You can do something like this (written off the top of my head so there might be bugs, you can compare against our other initializers): 

```javascript
export default {
  name: 'android-service-work',
  after: 'sniff-capabilities',
  initialize(container) {
    const caps = container.lookup('capabilities-main');
    if (caps.isAndroid) {
      // initialize service work
    }
  }
}
```

**Comment by coding-horror** [comment]
This should wait for 1.6 before merge

**Comment by coding-horror** [comment]
How do you feel about this now @eviltrout

**Comment by eviltrout** [comment]
I'd still prefer this to only execute on android before merging.

**Comment by xfalcox** [comment]
Did the changes :fireworks:

File: app/assets/javascripts/discourse/initializers/android-app-banner-service-worker.js.es6 [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+  initialize(container) {
+    const caps = container.lookup('capabilities:main');
+    if (caps.isAndroid) {
+      if ('serviceWorker' in navigator) {
+        navigator.serviceWorker.register(Discourse.BaseUri + '/service-worker.js', {scope: './'})
+          .then(function (r) {
+            // Service Worker Success
+          })
+          .catch(function (whut) {
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/initializers/android-app-banner-service-worker.js.es6]: Perhaps rename `whut` to just `e` or `error`?

Comment (tgxworld) [inline app/assets/javascripts/discourse/initializers/android-app-banner-service-worker.js.es6]: Actually if we're not doing anything with the response, we should leave out the calls. `then` and `catch`

**Comment by xfalcox** [comment]
@tgxworld removed the unnecessary bits, and added a test for https, since it obligatory.

**Comment by tgxworld** [comment]
@xfalcox Thanks! The PR needs a rebase :)

**Comment by xfalcox** [comment]
Okay, I think it's rebased :smile:

**Comment by xfalcox** [comment]
I guess that

**Comment by eviltrout** [comment]
Actually we don't need to ask people to rebase anymore @tgxworld. When
merging github now has an option to squash commits :)

**Comment by xfalcox** [comment]
Ruby 2.0 build failed, but I think it's unrelated.

**Comment by tgxworld** [comment]
@eviltrout but we'll still need to rebase if there is a merge conflict.

**Comment by xfalcox** [comment]
Hey @tgxworld  anything else?

**Comment by coding-horror** [comment]
if this looks OK to you can we merge it @eviltrout ?

**Comment by eviltrout** [comment]
Yes it looks good. Thanks for the effort and back and forth @xfalcox !

**Comment by xfalcox** [comment]
Thanks everyone, it's already working on meta! :fireworks:

If you guys wanna test faster there is a flag on chrome://flags   #bypass-app-banner-engagement-checks

**Pull request merged**